src/ray/gcs/store_client/redis_store_client.cc

```diff
@@ -38,29 +38,21 @@ Status RedisStoreClient::AsyncPutWithIndex(const std::string &table_name,
                                            const std::string &index_key,
                                            const std::string &data,
                                            const StatusCallback &callback) {
-  auto write_callback = [this, table_name, key, data, callback](Status status) {
-    if (!status.ok()) {
-      // Run callback if failed.
-      if (callback != nullptr) {
-        callback(status);
-      }
-      return;
-    }
-
-    // Write data to Redis.
-    status = DoPut(GenRedisKey(table_name, key), data, callback);
-
-    if (!status.ok()) {
-      // Run callback if failed.
-      if (callback != nullptr) {
-        callback(status);
-      }
-    }
-  };
-
+  // NOTE: To ensure the atomicity of `AsyncPutWithIndex`, we can't write data to Redis in
+  // the callback function of index writing.
   // Write index to Redis.
-  std::string index_table_key = GenRedisKey(table_name, key, index_key);
-  return DoPut(index_table_key, key, write_callback);
+  const auto &index_table_key = GenRedisKey(table_name, key, index_key);
+  RAY_CHECK_OK(DoPut(index_table_key, key, nullptr));
+
+  // Write data to Redis.
+  const auto &status = DoPut(GenRedisKey(table_name, key), data, callback);
```

Comment:
It is better to add some comment here to explain why it's ok to move the second `DoPut` out from the first ones callback.

Reply:
Good idea, thanks.